### PR TITLE
Adjust runc to new opencontainers/specs version

### PR DESCRIPTION
**Pull request by LK4D4**
I deleted possibility to specify config file from commands for now.
Until we decide how it'll be done. Also I changed runc spec interface to
write config files instead of output them.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,16 +4,17 @@
...
 
 ### OCF
 
-How does `runc` integrate with the Open Container Format?  `runc` depends on the types
-specified in the [specs](https://github.com/opencontainers/specs) repository.  Whenever
-the specification is updated and ready to be versioned `runc` will update it's dependency
+How does `runc` integrate with the Open Container Initiative Specification?
+`runc` depends on the types specified in the
+[specs](https://github.com/opencontainers/specs) repository. Whenever the
+specification is updated and ready to be versioned `runc` will update it's dependency
```
Comment (mrunalp) [inline README.md]: s/it's/its

File: spec.go [hunk 2]
```diff
@@ -214,23 +263,34 @@ var namespaceMapping = map[string]configs.NamespaceType{
 
 // loadSpec loads the specification from the provided path.
 // If the path is empty then the default path will be "config.json"
-func loadSpec(path string) (*specs.LinuxSpec, error) {
-	if path == "" {
-		path = "config.json"
-	}
+func loadSpec() (spec *specs.LinuxSpec, rspec *specs.LinuxRuntimeSpec, err error) {
+	path := "config.json"
```
Comment (mrunalp) [inline spec.go]: path --> cpath or configPath ?

**Comment by LK4D4** [comment]
@mrunalp fixed

File: README.md [hunk 3]
```diff
@@ -280,8 +352,8 @@ To test using Docker's `busybox` image follow these steps:
 mkdir rootfs
 tar -C rootfs -xf busybox.tar
 ```
-* Create a file called `config.json` using the example from above.  You can also
-generate a spec using `runc spec`, redirecting the output into `config.json`
+* Create a files `config.json` and `runtime.json` using the example from above.  You can also
```
Comment (mrunalp) [inline README.md]: s/a files//
or s/a/the/

**Comment by LK4D4** [comment]
@mrunalp fixed

File: README.md [hunk 4]
```diff
@@ -41,229 +42,300 @@ PID   USER     COMMAND
...
-/ $
-```
-Note: the use of the `start` command is required when specifying a 
-configuration file.
+### OCI Container JSON Format:
 
-### OCF Container JSON Format:
-
-Below is a sample `config.json` configuration file. It assumes that
+Below are sample `config.json` and `runtime.json` configuration files. It assumes that
```
Comment (wking) [inline README.md]: I think it makes sense to move this generic OCS documentation into opencontainers/specs.  Then you can link to that overview from opencontainers/runc, and not worry about keeping your OCS docs in sync with the upstream spec.  Folks suggesting changes to the spec would just update the docs when PRing their changes, and the OCS maintainers would review the whole change as a cohesive unit.

File: README.md [hunk 5]
```diff
@@ -280,8 +352,8 @@ To test using Docker's `busybox` image follow these steps:
 mkdir rootfs
 tar -C rootfs -xf busybox.tar
 ```
-* Create a file called `config.json` using the example from above.  You can also
-generate a spec using `runc spec`, redirecting the output into `config.json`
+* Create `config.json` and `runtime.json` using the example from above.  You can also
+generate a spec using `runc spec`, which will create those files for you.
```
Comment (wking) [inline README.md]: This, on the other hand, is runC-specific information that should stay in this repository.

File: README.md (hunk 4 above)
Comment (LK4D4) [inline README.md]: Agree, let's do this in another PR though. It was supposed to be mostly code PR to get runc working.

Comment (wking) [inline README.md]: On Tue, Sep 01, 2015 at 04:29:17PM -0700, Alexander Morozov wrote:
> > -### OCF Container JSON Format:
> > -
> > -Below is a sample `config.json` configuration file. It assumes that
> > +Below are sample `config.json` and `runtime.json` configuration files. It assumes that
> 
> Agree, let's do this in another PR though. It was supposed to be
> mostly code PR to get runc working.

Sounds good.

**Comment by crosbymichael** [comment]
After looking at the configs in this PR I don't see how mounts in the config.json with Name are related to mounts in the RuntimeSpec?

**Comment by mrunalp** [comment]
Yeah, need a check to make sure that all the mounts in config.json are satisfied in runtime.json.

**Comment by crosbymichael** [comment]
@mrunalp i think it's a problem in the spec.  The Name is not used and I don't see how they can even relate to eachother

**Comment by wking** [comment]
On Tue, Sep 01, 2015 at 04:58:23PM -0700, Michael Crosby wrote:
> @mrunalp i think it's a problem in the spec.  The Name is not used
> and I don't see how they can even relate to eachother

I agree that it's a spec problem.  See discussion yesterday on IRC
between @mrunalp, @lk4d4, and myself.

**Comment by crosbymichael** [comment]
I just think the `Mount` struct is missing `Name string` and that should fix the issues.

**Comment by mrunalp** [comment]
@crosbymichael I agree.

**Comment by LK4D4** [comment]
They related Path <-> Destination. If we'll have name in mounts, then Path is redundant. And then mounts probably should be a map.

**Comment by LK4D4** [comment]
@crosbymichael 338 line in spec.go

**Comment by mrunalp** [comment]
I think that it makes sense to add Name and take out destination.

**Comment by wking** [comment]
On Tue, Sep 01, 2015 at 06:25:10PM -0700, Alexander Morozov wrote:
> They related Path <-> Destination. If we'll have name in mounts,
> then Path is redundant. And then mounts probably should be a map.

I agree that we should use maps on both ends, and the current
Mount.Destination and MountPoint.Name should become keys to those
maps.  Shall I make an opencontainers/specs PR to that effect, or does
someone else want to handle it?

**Comment by LK4D4** [comment]
@wking I'll do it tomorrow. I'm working on docker integration and probably will notice more rough edges in implementation.

**Comment by wking** [comment]
On Tue, Sep 01, 2015 at 08:26:12PM -0700, Alexander Morozov wrote:
> I'll do it tomorrow…

Ok.  If we went with two maps, we'd need logic in the runtime to order
the mounts, so it's probably easier to keep the MountPoints entry as
an array (because they know about the destinations) and turn the
Mounts entry into a map (so you can't duplicate keys).  In fact, it
might make sense to allow duplicate Names in the MountPoints array (in
which case the same entry from the Mounts map would be mounted at both
destinations.

File: README.md [hunk 6]
```diff
@@ -41,229 +42,300 @@ PID   USER     COMMAND
...
-                        "major": 1,
-                        "minor": 9,
-                        "permissions": "rwm",
-                        "fileMode": 438,
-                        "uid": 0,
-                        "gid": 0
-                }
-        ],
-    }
+	"mounts": [
```
Comment (crosbymichael) [inline README.md]: I think this example needs to be updated.

Comment (LK4D4) [inline README.md]: You're right. update in a second

File: spec.go [hunk 7]
```diff
@@ -22,6 +23,10 @@ import (
 var specCommand = cli.Command{
 	Name:  "spec",
 	Usage: "create a new specification file",
+	Flags: []cli.Flag{
+		cli.StringFlag{Name: "config-output, c", Value: "config.json", Usage: "path to spec file for writing"},
```
Comment (crosbymichael) [inline spec.go]: maybe just `config-file` and `runtime-file` ?

Comment (LK4D4) [inline spec.go]: makes sense for me

**Comment by LK4D4** [comment]
@crosbymichael fixed

File: README.md [hunk 8]
```diff
@@ -41,229 +42,293 @@ PID   USER     COMMAND
...
+			"source": "cgroup",
+			"options": [
+				"nosuid",
+				"noexec",
+				"nodev",
+				"relatime",
+				"ro"
+			]
+		}
+	],
```
Comment (mrunalp) [inline README.md]: Is that an additional ] ?

Comment (LK4D4) [inline README.md]: looks like it

**Comment by LK4D4** [comment]
@mrunalp fixed bracket.

**Comment by mrunalp** [comment]
LGTM

**Comment by LK4D4** [comment]
ping @crosbymichael

**Comment by crosbymichael** [comment]
For the usability of the `runc specs` command.  It will overwrite your existing files without you knowing.  Maybe you should pass a `-f` or it will not write the file if it already exists?

**Comment by crosbymichael** [comment]
I get a panic if I delete the `runtime.json` file and try to run the container. 


```
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x40931c]

goroutine 1 [running]:
main.createLibcontainerConfig(0xc8200b0bb0, 0x5, 0xc8200b8200, 0x0, 0x5, 0x0, 0x0)
        /home/michael/development/gocode/src/github.com/opencontainers/runc/spec.go:314 +0x139c
main.startContainer(0xc8200d6000, 0xc8200b8200, 0x0, 0x0, 0x0, 0x0)
        /home/michael/development/gocode/src/github.com/opencontainers/runc/start.go:67 +0xa8
main.glob.func9(0xc8200d6000)
        /home/michael/development/gocode/src/github.com/opencontainers/runc/start.go:44 +0x2d1
github.com/codegangsta/cli.(*App).Run(0xc8200c8000, 0xc82000a2e0, 0x1, 0x1, 0x0, 0x0)
        /home/michael/development/gocode/src/github.com/opencontainers/runc/Godeps/_workspace/src/github.com/codegangsta/cli/app.go:159 +0xbe7
main.main()
        /home/michael/development/gocode/src/github.com/opencontainers/runc/main.go:92 +0x858

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1696 +0x1

goroutine 18 [syscall]:
os/signal.loop()
        /usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
        /usr/local/go/src/os/signal/signal_unix.go:28 +0x37

```

**Comment by cyphar** [comment]
Can we bump the specs dependancy further to include opencontainers/specs@834fb5db52bc398511885f7fe9a7a996fd96d2c3 so we can move forward with #58?

**Comment by cyphar** [comment]
@LK4D4 Can we bump the specs dependency further to include opencontainers/specs@834fb5db52bc398511885f7fe9a7a996fd96d2c3 so we can move forward with #58?

**Comment by LK4D4** [comment]
@cyphar sorry, missed comment, yes, I'll update

**Comment by LK4D4** [comment]
@crosbymichael I made `-config-file` and `-runtime-file` mandatory. As well as existing `runtime.json` for `runc start`.

**Comment by crosbymichael** [comment]
@LK4D4 i cannot run the `runc specs` command anymore.

```bash
runc specs
JSON runtime config file at runtime.json not found
```

**Comment by LK4D4** [comment]
@crosbymichael hehe, it's `spec` :)

**Comment by crosbymichael** [comment]
Getting error when running a container with default configs:

```bash
WARN[0000] signal: killed                               
FATA[0000] Container start failed: [8] System error: mountpoint for cgroup not found 

```

**Comment by LK4D4** [comment]
@crosbymichael Probably same problem as for #259
(works on my machine, btw)

**Comment by crosbymichael** [comment]
@LK4D4 works on master, build this new version broken

**Comment by LK4D4** [comment]
@crosbymichael For me it's same on master. #259 fixes it.

**Comment by mrunalp** [comment]
yeah that has been fixed by #259. Maybe rebase here will help testing.

**Comment by LK4D4** [comment]
@crosbymichael @mrunalp rebased

**Comment by crosbymichael** [comment]
@LK4D4 ok, it is working now

**Comment by crosbymichael** [comment]
@LK4D4 the `runc spec` command is kinda confusing now because i have to specify the flags but on `start` the names cannot be changed from the default.  Maybe the flags should have the default names in them, it just will error out if the file already exists.  Also maybe `start` should have the same flags for specifying the files?

**Comment by wking** [comment]
On Thu, Sep 10, 2015 at 02:15:31PM -0700, Michael Crosby wrote:
> @LK4D4 the `runc spec` command is kinda confusing now because i have
> to specify the flags but on `start` the names cannot be changed from
> the default.

The initial position of this PR was not to allow alternative names
[1].  I'm personally in favor of having the names be configurable, but
I think the “punt for now and just hard-code the names” is a
reasonable position for this PR.  So that would be “drop the flags
from ‘runc spec’”.

[1]: https://github.com/opencontainers/runc/pull/242#issue-104305897

**Comment by LK4D4** [comment]
@crosbymichael I added flags everywhere and made spec error out if files exist.

**Comment by LK4D4** [comment]
ping @crosbymichael @mrunalp

File: spec.go [hunk 9]
```diff
@@ -22,6 +23,10 @@ import (
 var specCommand = cli.Command{
 	Name:  "spec",
 	Usage: "create a new specification file",
+	Flags: []cli.Flag{
+		cli.StringFlag{Name: "config-file, c", Value: "config.json", Usage: "path to spec file for writing"},
+		cli.StringFlag{Name: "runtime-file, r", Value: "runtime.json", Usage: "path for runtime file for writing"},
```
Comment (mrunalp) [inline spec.go]: path to runtime file

**Comment by crosbymichael** [comment]
@LK4D4 is this updated with the 0.0.1 spec?  Maybe we should also bump the runc version up one with these chagnes?

File: spec.go [hunk 10]
```diff
@@ -195,42 +228,75 @@ var specCommand = cli.Command{
 				},
 			},
 		}
+		checkNoFile := func(name string) error {
+			_, err := os.Stat(name)
+			if err == nil {
+				return fmt.Errorf("File %s exist. Remove it first", name)
```
Comment (mrunalp) [inline spec.go]: s/exist/exists

**Comment by LK4D4** [comment]
@crosbymichael okay

**Comment by LK4D4** [comment]
Who knows how am I supposed to convert rlimit.Type to int now?

**Comment by mrunalp** [comment]
@LK4D4 If you pulled the latest spec changes, then we need a conversion similar to capabilities.

**Comment by LK4D4** [comment]
@mrunalp uh, oh. okay, then probably today not best day to do it. btw it's wrong for capabilities as well :/

**Comment by mrunalp** [comment]
From /usr/include/bits/resource.h

```
  RLIMIT_CPU = 0,
  RLIMIT_FSIZE = 1,
  RLIMIT_DATA = 2,
  RLIMIT_STACK = 3,
  RLIMIT_CORE = 4,
  RLIMIT_RSS = 5,
  RLIMIT_NPROC = 6,
  RLIMIT_NOFILE = 7,
  RLIMIT_MEMLOCK = 8,
  RLIMIT_AS = 9,
  RLIMIT_LOCKS = 10,
  RLIMIT_SIGPENDING = 11,
  RLIMIT_MSGQUEUE = 12,
  RLIMIT_NICE = 13,
  RLIMIT_RTPRIO = 14,
  RLIMIT_RTTIME = 15,
  RLIMIT_NLIMITS = 16,
```

**Comment by mrunalp** [comment]
@LK4D4 You can leave it out for now if you want. I can fix it up in another PR.

**Comment by mrunalp** [comment]
You would have to use an older version of the spec though.

**Comment by LK4D4** [comment]
@mrunalp so, cool. There is no NPROC in syscall package. And I wonder if those constants different between arches. I think I can just do it at monday or tuesday. There is no need to hurry.

**Comment by mrunalp** [comment]
@LK4D4 sounds good.

**Comment by LK4D4** [comment]
@mrunalp do you know difference between `RLIMIT_` constants from glibc and kernel? Because they're like different.

**Comment by mrunalp** [comment]
@LK4D4 http://www.gnu.org/software/libc/manual/html_node/Limits-on-Resources.html Looks like some values aren't valid on BSDs.

**Comment by cyphar** [comment]
The values change between some architectures. If you compare `arch/alpha/include/uapi/asm/resource.h`, `arch/mips/include/uapi/asm/resource.h` and `arch/sparc/include/uapi/asm/resource.h` you get different values. `include/uapi/asm-generic/resource.h`then provides default values.

However, for the architectures Go supports, these values appear to be constant.

**Comment by LK4D4** [comment]
@cyphar Yup, I think I'll just take generic for now.

**Comment by LK4D4** [comment]
@mrunalp @crosbymichael I updated spec

**Comment by mrunalp** [comment]
@LK4D4 
```
[root@localhost busybox]# ls
bin   config.json  etc   lib    linuxrc  mnt       opt   root  runtime.json  sys  usr
code  dev          home  lib64  media    mountrun  proc  run   sbin          tmp  var
[root@localhost busybox]# runc
JSON specification file at  not found
[root@localhost busybox]# runc config.json 
JSON specification file at  not found
[root@localhost busybox]# 
```

**Comment by LK4D4** [comment]
@mrunalp It sorta hurts me that `runc` is alias to `runc start`. How about to remove this alias?

**Comment by mrunalp** [comment]
@LK4D4 If nothing is specified, we can look in the current directory for both the files like before. I am agnostic to the start alias.

**Comment by LK4D4** [comment]
@mrunalp We're looking for them, but for `start` command :)

**Comment by mrunalp** [comment]
@LK4D4 Hmmm

**Comment by wking** [comment]
On Mon, Sep 14, 2015 at 12:38:16PM -0700, Alexander Morozov wrote:
> @mrunalp It sorta hurts me that `runc` is alias to `runc start`. How
> about to remove this alias?

Previous discussion about this alias in #210 (after [1]).  Maybe a
deprecation phase where you keep the alias but warn to stderr if folks
use it?  Then remove the alias after a few months of that?

[1]: https://github.com/opencontainers/runc/pull/210#discussion_r37681561

**Comment by LK4D4** [comment]
@mrunalp I made `-config-file` and `-runtime-file` global flags, now `runc` should work.

**Comment by crosbymichael** [comment]
@LK4D4 we can remove the alias now since this is a big update with the multiple spec files so it's breaking anyways

**Comment by LK4D4** [comment]
@crosbymichael I already made flags global, so alias works. Not sure if you still want to remove it.

**Comment by crosbymichael** [comment]
@LK4D4 i would probably do it now since we are making large breaking changes anyways

**Comment by mrunalp** [comment]
Sounds good to me. :+1:

**Comment by wking** [comment]
On Mon, Sep 14, 2015 at 01:22:35PM -0700, Michael Crosby wrote:
> @LK4D4 we can remove the alias now since this is a big update with
> the multiple spec files so it's breaking anyways

Updating the spec breaks the bundle-author API (which runC does
frequently).  Dropping the ‘runc’ → ‘runc start’ alias breaks the
runC-caller API (which runC hasn't done yet, at least for the bare
‘runc’ invocation).  Still, I don't foresee much trouble with breaking
the runC-caller API, I just think it would be gentler to give folks a
window to adjust.  But we can sort out things like that once we start
versioning runC ;).

**Comment by LK4D4** [comment]
@crosbymichael I dropped alias. Now it's only `runc start`.

**Comment by crosbymichael** [comment]
@LK4D4 ok but i don't think u pushed the commit

**Comment by LK4D4** [comment]
@crosbymichael Pushed. Sorry for that.

**Comment by LK4D4** [comment]
@crosbymichael I think now I pushed something.

**Comment by mrunalp** [comment]
Looks fine to me but janky isn't working

File: restore.go [hunk 11]
```diff
@@ -18,23 +18,53 @@ var restoreCommand = cli.Command{
...
+			Usage: "path for runtime file for writing",
+		},
 	},
 	Action: func(context *cli.Context) {
 		imagePath := context.String("image-path")
 		if imagePath == "" {
 			imagePath = getDefaultImagePath(context)
 		}
-		spec, err := loadSpec(context.Args().First())
+		spec, rspec, err := loadSpec(context.GlobalString("config-file"), context.GlobalString("runtime-file"))
```
Comment (crosbymichael) [inline restore.go]: This should not be `GlobalString` here and just be `String`

**Comment by crosbymichael** [comment]
I think maybe these changes are needed:

```patch
diff --git a/main.go b/main.go
index 5f59783..33a86fc 100644
--- a/main.go
+++ b/main.go
@@ -5,10 +5,11 @@ import (
 
 	"github.com/Sirupsen/logrus"
 	"github.com/codegangsta/cli"
+	"github.com/opencontainers/specs"
 )
 
 const (
-	version = "0.2"
+	version = "0.3"
 	usage   = `Open Container Initiative runtime
 
 runc is a command line client for running applications packaged according to
@@ -54,7 +55,7 @@ func main() {
 		},
 		cli.StringFlag{
 			Name:  "root",
-			Value: "/run/oci",
+			Value: specs.LinuxStateDirectory,
 			Usage: "root directory for storage of container state (this should be located in tmpfs)",
 		},
 		cli.StringFlag{
diff --git a/restore.go b/restore.go
index 094a4d3..bf12cb4 100644
--- a/restore.go
+++ b/restore.go
@@ -60,7 +60,7 @@ var restoreCommand = cli.Command{
 		if imagePath == "" {
 			imagePath = getDefaultImagePath(context)
 		}
-		spec, rspec, err := loadSpec(context.GlobalString("config-file"), context.GlobalString("runtime-file"))
+		spec, rspec, err := loadSpec(context.String("config-file"), context.String("runtime-file"))
 		if err != nil {
 			fatal(err)
 		}
```

**Comment by LK4D4** [comment]
@crosbymichael Incorporated your changes.

**Comment by crosbymichael** [comment]
LGTM

**Comment by mrunalp** [comment]
LGTM
